app/scripts/controllers/network/network-controller.js

```diff
@@ -476,4 +502,116 @@ export default class NetworkController extends EventEmitter {
     this._provider = provider;
     this._blockTracker = blockTracker;
   }
+
+  /**
+   * Network Configuration management functions
+   */
+
+  /**
+   * Adds a network configuration if the rpcUrl is not already present on an
+   * existing network configuration. Otherwise updates the entry with the matching rpcUrl.
+   *
+   * @param {NetworkConfiguration} networkConfiguration - The network configuration to add or, if rpcUrl matches an existing entry, to modify.
+   * @param {object} options
+   * @param {boolean} options.setActive - An option to set the newly added networkConfiguration as the active provider.
+   * @param {string} options.source - Where the upsertNetwork event originated - used for event metrics.
```

Comment:
Could we split this into two properties - referrer and source? Combining them into two seems error prone. Now we're forced to keep this list here of all valid internal sources, and we don't get any type safety for this option.

This has already resulted in a bug; the `referrer.url` property is being set to `source` directly here when it should be `metamask` for internally-added networks.

Reply:
Sure. I'm assuming you mean two arguments?

Reply:
Oh I misread this. Let me take another look

Reply:
Yep so I'm adding a `referrer` property to this options object.

Reply:
done here: https://github.com/MetaMask/metamask-extension/pull/17421/commits/ccfc91787ab01012269c2d2bd97fa9b3c0de5ace